magento/magento2#39959 catalog_product_save_before observer throws date-related error when using REST API without store-level values (getFinalPrice() issue)

KrasnoshchokBohdan opened the pull request:
Adjusted the processing of `SpecialFromDate` to ensure proper formatting when the date is provided as a `DateTimeInterface` instance. This prevents errors arising during `getFinalPrice()` execution under certain scenarios.

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
I focused on investigation error 
`TypeError: strtotime(): Argument #1 ($datetime) must be of type string, DateTime given in lib/internal/Magento/Framework/Stdlib/DateTime/Timezone.php:275` 
the reason for it must be 
\Magento\Catalog\Observer\SetSpecialPriceStartDate::execute
<img width="1132" alt="image" src="https://github.com/user-attachments/assets/7c0d35f0-c658-4c02-b4ab-3e7e41f31ee2" />


### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#[39959](https://github.com/magento/magento2/issues/39959)

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->

1. Create a custom module with an observer for catalog_product_save_before.
2. Inside the observer, call $product->getFinalPrice().
3. Save product from storeview scope from admin panel (check the video from issue description)
4. Save a product using REST API: POST /rest/default/V1/products
5. Use the following JSON payload (no store-level pricing):
<details>
  <summary>payload</summary>

```json
{
  "product": {
    "sku": "testing-special-date",
    "price": 120,
    "status": 0,
    "visibility": 4,
    "type_id": "simple",
    "extension_attributes": {
      "stock_item": {
        "stock_id": 1,
        "qty": 207,
        "is_in_stock": true
      }
    },
    "custom_attributes": [
      {
        "attribute_code": "tax_class_id",
        "value": "2"
      },
      {
        "attribute_code": "special_price",
        "value": "100.00"
      }
    ]
  },
  "saveOptions": true
}
```
</details>

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

KrasnoshchokBohdan committed:
magento/magento2#39959 catalog_product_save_before observer throws date-related error when using REST API without store-level values (getFinalPrice() issue)

Adjusted the processing of `SpecialFromDate` to ensure proper formatting when the date is provided as a `DateTimeInterface` instance. This prevents errors arising during `getFinalPrice()` execution under certain scenarios.

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39959

KrasnoshchokBohdan committed:
magento/magento2#39959 catalog_product_save_before observer throws date-related error when using REST API without store-level values (getFinalPrice() issue)

Updated special price logic to ensure date formats are handled consistently when calculating and setting the "special from" date. This improves compatibility and prevents potential type issues with date-related operations.

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39959

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39959 catalog_product_save_before observer throws date-related error when using REST API without store-level values (getFinalPrice() issue)

- fix static test errors

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39959

KrasnoshchokBohdan commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39959

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for the contribution!

The changes seems good to us, but some suggestion as follows:
- The PR changes can be covered by unit and integration tests. A test should verify that setting a special price via API or admin behaves correctly.
- Add inline comments to explain the rationale behind formatting `special_from_date`.

KrasnoshchokBohdan committed:
magento/magento2#39959 catalog_product_save_before observer throws date-related error when using REST API without store-level values (getFinalPrice() issue)

Add tests and logic for handling special_from_date updates

Introduced integration and unit tests to validate behavior of the special price start date logic. Ensures special_from_date is set automatically if not provided and remains unchanged when already set. Also added comments and adjustments to observer logic to handle special price updates consistently.

KrasnoshchokBohdan commented:
@magento run all tests

engcom-Dash commented:
@magento run Integration Tests

engcom-Hotel commented:
@magento run Functional Tests B2B, Functional Tests CE

engcom-Dash committed:
Merge branch '2.4-develop' into fix-for-issue-39959

engcom-Dash commented:
@magento run all tests

engcom-Dash committed:
Merge branch '2.4-develop' into fix-for-issue-39959

engcom-Dash commented:
@magento run all tests

engcom-Dash commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE
